package/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+    #[serde(default)]
+    pub path: PathBuf,
+}
+
+impl GitDependency {
+    /// If this git dependency specifies a relative path, make it absolute.
+    pub fn relative_to(&self, relative_to: Option<&Path>) -> Result<Self, Error> {
+        if self.url.scheme.as_str() == "file" {
+            // unwrap: the url ultimately came from a nickel file, which is always valid UTF-8.
+            let path = Path::new(self.url.path.to_str().unwrap());
+            if path.is_relative() {
+                match relative_to {
+                    Some(relative_to) => {
+                        let abs_path = relative_to.join(path);
+                        Ok(GitDependency {
+                            url: Url::try_from(abs_path.as_path()).map_err(|e| {
+                                Error::InvalidUrl {
+                                    url: abs_path.display().to_string(),
+                                    msg: e.to_string(),
+                                }
+                            })?,
+                            ..self.clone()
+                        })
+                    }
+                    None => Err(Error::RelativeGitImport {
+                        path: path.to_owned(),
+                    }),
+                }
+            } else {
+                Ok(self.clone())
+            }
+        } else {
+            Ok(self.clone())
+        }
+    }
+}
+
+/// A source includes the place to fetch a package from (e.g. git or a registry),
+/// along with possibly some narrowing-down of the allowed versions (e.g. a range
+/// of versions, or a git commit id).
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Deserialize)]
+pub enum Dependency {
+    Git(GitDependency),
+    Path { path: PathBuf },
+}
+
+impl Dependency {
+    pub fn matches(&self, entry: &LockFileDep) -> bool {
+        match self {
+            Dependency::Git(git) => match &entry.spec {
+                Some(locked_git) => git == locked_git,
+                None => false,
+            },
+            Dependency::Path { .. } => true,
+        }
+    }
+}
+
+mod serde_url {
+    use serde::{de::Error, Deserialize, Serialize as _};
+
+    pub fn serialize<S: serde::Serializer>(url: &gix::Url, ser: S) -> Result<S::Ok, S::Error> {
+        // unwrap: locked urls can only come from nickel strings in the manifest file, which must be
+        // valid utf-8
+        std::str::from_utf8(url.to_bstring().as_slice())
+            .unwrap()
+            .serialize(ser)
+    }
+
+    pub fn deserialize<'de, D: serde::Deserializer<'de>>(de: D) -> Result<gix::Url, D::Error> {
+        let s = String::deserialize(de)?;
+        gix::Url::try_from(s).map_err(|e| D::Error::custom(e.to_string()))
+    }
+}
+
+/// A precise package version, in a format suitable for putting into a lockfile.
+#[serde_with::serde_as]
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize, PartialOrd, Ord)]
+pub enum Precise {
```

Comment:
Nitpick, feel free to ignore: `Precise` is quite general. Is it a `PreciseDep` ? `PreciseGitDep` ?

Reply:
I went with `PrecisePkg`. (I think "package" is more than "dependency" here because this can be used to precisely specify a package in general, not only when we're talking about a package that's a dependency of another package.)